File: Exercises2024/Chapter06_vle_free_energy/C/1-Gibbs/Source/gibbs.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include "system_2.h"
#include "parameter.h"
#include "ran_uniform.h"

double X[Npmax],Y[Npmax],Z[Npmax];
int Id[Npmax],Npart,Npbox[2];
double Box[2],Hbox[2],Temp,Beta;
double  Eps4,Eps48,Sig2,Mass,Rc[2],Rc2[2];
double Chp[2];
int Ichp[2]; 

int main()
{
  //    ---Gibbs-Ensemble Simulation Of The Lennard-Joned Fluid
  int  Equil, Prod, Nsamp, I, Icycl, Ndispl, Attempt; 
  int  Nacc, Ncycl, Nmoves, Imove, Nvol, AcceptVolume; 
  int  AttemptVolume, BoxID, Nswap, AcceptSwap, AttemptSwap;
  double En[2], Ent[2], Vir[2], Virt[2], Vmax, Dr; 
  double Ran, Succ;
  FILE* Fileptr;
  
  // initialize the random number generator with the system time
  InitializeRandomNumberGenerator(time(0l));
  
  printf("**************** GIBBS ***************\n");
  
  //    ---Initialize System
  Readdat(&Equil, &Prod, &Nsamp, &Ndispl, &Dr, &Nvol, &Vmax, &Nswap, &Succ);
  Nmoves = Ndispl + Nvol + Nswap;
  
  //    ---Total Energy Of The System
  for(BoxID=0;BoxID<2;BoxID++)
    {
      Toterg(&(En[BoxID]),&(Vir[BoxID]),BoxID);
      printf("Box: %d\n",BoxID);
      printf("Total Energy Initial Configuration: %lf\n",En[BoxID]);
      printf("Total Virial Initial Configuration: %lf\n",Vir[BoxID]);
    }

  //    ---Start MC Cycle
  
  for(I=1;I<3;I++)
    {
      //       --- I=1 Equilibration
      //       --- I=2 Production
      
      if(I==1)
	{
	  Ncycl = Equil;
	  if(Ncycl!=0) printf("Start Equilibration\n"); 
	}
      else
	{
	  Ncycl = Prod;
	  if(Ncycl!=0) printf("Start Production\n"); 
	}
      Attempt = 0;
      Nacc = 0;
      AttemptVolume = 0;
      AcceptVolume = 0;
      AttemptSwap = 0;
      AcceptSwap = 0;
      
      //       ---Initialize Calculation Chemical Potential
      
      Init_Chem(0);
      
      //       ---Intialize The Subroutine That Adjust The Maximum Displacement

      Adjust(Attempt, Nacc, Dr, AttemptVolume, AcceptVolume, Vmax, Succ);
      
      for(Icycl=0;Icycl<Ncycl;Icycl++)
	{
	  for(Imove=0;Imove<Nmoves;Imove++)
            {
	      Ran = RandomNumber()*(Ndispl+Nvol+Nswap);
	      if(Ran<(double)(Ndispl)) 
		{ 
		  //                ---Attempt To Displace A Particle
                  Mcmove(En, Vir, &Attempt, &Nacc, Dr);
		}
	      else if (Ran<(double)(Ndispl+Nvol))
		{
		  //                ---Attempt To Change The Volume
                  Mcvol(En, Vir, &AttemptVolume,&AcceptVolume, Vmax);
		}
	      else
		{
		  //                ---Attemp To Exchange Particles
		  Mcswap(En, Vir, &AttemptSwap, &AcceptSwap);
		}
            }
	  if(I==2)
            {  
	      //             ---Sample Averages
	      if((Icycl%Nsamp)==0) Sample(Icycl, En, Vir);
            }
	  if((Icycl%(Ncycl/10))==0)
            {
	      printf("======>> Done %d out of %d %d %d\n", Icycl, Ncycl,
		     Npbox[0], Npbox[1]);
	      
	      //             ---Write Intermediate Configuration To File
	      Fileptr=fopen("fort.8","w");
	      Store(Fileptr, Dr, Vmax);
	      fclose(Fileptr);
	      for(BoxID=0;BoxID<2;BoxID++)
		{
		  printf("Density: %g\n",(double)Npbox[BoxID]/pow(Box[BoxID],3.));
		}
	      //             ---Adjust Maximum Displacements
	      
	      Adjust(Attempt, Nacc, Dr, AttemptVolume, AcceptVolume, Vmax, Succ);
            }
	}
      if(Ncycl!=0)
	{
	  if(Attempt!=0) 
	    {
	      printf("Number of Att. to Displ. a Part.: %d\n", Attempt);
	      printf("Success: %d (= %lf %%)\n",Nacc,
		     100.*(double)(Nacc)/(double)(Attempt));
            }
	  if(AttemptVolume!=0) 
            {
	      printf("Number of Att. to Change Volume: %d\n", AttemptVolume);
	      printf("Success: %d (= %lf %%)\n",AcceptVolume,
		     100.*(double)(AcceptVolume)/(double)(AttemptVolume));
            }
	  if(AttemptSwap!=0) 
            {
	      printf("Number of Att. to Exchange Part.: %d\n", AttemptSwap);
	      printf("Success: %d (= %lf %%)\n",AcceptSwap,
		     100.*(double)(AcceptSwap)/(double)(AttemptSwap));
            }
	  for(BoxID=0;BoxID<2;BoxID++)
            { 
	      //             ---Test Total Energy
	      Toterg(&(Ent[BoxID]), &(Virt[BoxID]), BoxID);
	      if(fabs(Ent[BoxID]-En[BoxID])>1e-6) 
		{
                  printf(" ######### Problems Energy ################ \n");
		}
	      if(fabs(Virt[BoxID]-Vir[BoxID])>1e-6)
		{
                  printf(" ######### Problems Virial ################ \n");
		}
	      printf("Box: %d\n",BoxID);
	      printf("Total Energy End of Simulation: %g\n",Ent[BoxID]);
	      printf("Running Energy: %g\n",En[BoxID]);
	      printf("Difference: %g\n",(Ent[BoxID] - En[BoxID]));
	      printf("Total Virial End of Simulation: %g\n",Virt[BoxID]);
	      printf("Running Virial: %g\n",Vir[BoxID]);
	      printf("Difference: %g\n",(Virt[BoxID] - Vir[BoxID]));
	      printf("Number of particles: %d\n",Npbox[BoxID]);
	      printf("Volume of box: %g\n",pow(Box[BoxID],3.));
	      printf("Density: %g\n",(double)Npbox[BoxID]/pow(Box[BoxID],3.));
            }
	  //          ---Calculation Chemical Potential
	  Init_Chem(2);
	}
    }
  Fileptr=fopen("fort.21","w");
  Store(Fileptr, Dr, Vmax);
  fclose(Fileptr);
  return(0);
}

File: Exercises2024/Chapter06_vle_free_energy/Fortran/1-Gibbs/Source/gibbs.f
      Program Gibbs
      Implicit None

C     ---Gibbs-Ensemble Simulation Of The Lennard-Joned Fluid

      Integer Equil, Prod, Nsamp, I, Icycl, Ndispl, Attempt, 
     &        Nacc, Ncycl, Nmoves, Imove, Nvol, AcceptVolume, AttemptVolume, BoxID, Nswap, 
     &        AcceptSwap, AttemptSwap, Sstmm
      Double Precision En(2), Ent(2), Vir(2), Virt(2), Vmax, Dr, 
     &                 Ran, Succ,RandomNumber,M1

      Include 'parameter.inc'
      Include 'chem.inc'
      Include 'conf.inc'
      Include 'potential.inc'
      Include 'system.inc'
 
      M1 = 0.001d0*Dble(Mod((10+10*Sstmm()),1000))

      If(M1.Lt.0.001d0) M1 = 0.001d0
      If(M1.Gt.0.999d0) M1 = 0.999d0

      Call InitializeRandomNumberGenerator(M1)

      Write (6, *) '**************** GIBBS ***************'
      Write(66,'(7A20)')"#Time","En. box 1","En. box 2","Press. box 1",
     & "Press. box 2","Dens. (N/V) box 1", "Dens. (N/V) box 2"
      write(45,'(3A20)')"# Time","Dens. (N/V) box 1","Dens. (N/V) box 2"


C     ---Initialize Sysem

      Call Readdat(Equil, Prod, Nsamp, Ndispl, Dr, Nvol, Vmax, Nswap, 
     &             Succ)
      Nmoves = Ndispl + Nvol + Nswap

C     ---Total Energy Of The System

      Do BoxID = 1, 2
         Call Toterg(En(BoxID), Vir(BoxID), BoxID)
         Write (6, 99001) BoxID, En(BoxID), Vir(BoxID)
      End Do

C     ---Start Mc-Cycle

      Do I = 1, 2

C        --- I=1 Equilibration
C        --- I=2 Production

         If (I.Eq.1) Then
            Ncycl = Equil
            If (Ncycl.Ne.0) Write (6, *) ' Start Equilibration '
         Else
            If (Ncycl.Ne.0) Write (6, *) ' Start Production '
            Ncycl = Prod
         End If
         Attempt = 0
         Nacc = 0
         AttemptVolume = 0
         AcceptVolume = 0
         AttemptSwap = 0
         AcceptSwap = 0

C        ---Initialize Calculation Chemical Potential

         Call Init_Chem(0)

C        ---Intialize The Subroutine That Adjust The Maximum Displacement

         Call Adjust(Attempt, Nacc, Dr, AttemptVolume, AcceptVolume, Vmax, Succ)

         Do Icycl = 1, Ncycl
            Do Imove = 1, Nmoves
               Ran = RandomNumber()*Dble(Ndispl+Nvol+Nswap)
               If (Ran.Lt.Dble(Ndispl)) Then

C                 ---Attempt To Displace A Particle

                  Call Mcmove(En, Vir, Attempt, Nacc, Dr)
               Else If (Ran.Lt.Dble(Ndispl+Nvol)) Then

C                 ---Attempt To Change The Volume

                  Call Mcvol(En, Vir, AttemptVolume, AcceptVolume, Vmax)
               Else

C                 ---Attemp To Exchange Particles

                  Call Mcswap(En, Vir, AttemptSwap, AcceptSwap)
 
               End If
            End Do
            If (I.Eq.2) Then

C              ---Sample Averages

               If (Mod(Icycl,Nsamp).Eq.0) Call Sample(Icycl, En, Vir)
            End If
            If (Mod(Icycl,Ncycl/5).Eq.0) Then
               Write (6, *) '======>> Done ', Icycl, ' Out Of ', Ncycl, 
     &                      Npbox(1), Npbox(2)

C              ---Write Intermediate Configuration To File

               Call Store(8, Dr, Vmax)

C              ---Adjust Maximum Displacements

               Call Adjust(Attempt, Nacc, Dr, AttemptVolume, AcceptVolume, Vmax, Succ)
            End If
         End Do
         If (Ncycl.Ne.0) Then
            If (Attempt.Ne.0) Write (6, 99003) Attempt, Nacc, 
     &                               100.*Float(Nacc)/Float(Attempt)
            If (AttemptVolume.Ne.0) Write (6, 99004) AttemptVolume, AcceptVolume, 100.*Float(AcceptVolume)
     &                            /Float(AttemptVolume)
            If (AttemptSwap.Ne.0) Write (6, 99005) AttemptSwap, AcceptSwap, 
     &                             100.*Float(AcceptSwap)/Float(AttemptSwap)
            Do BoxID = 1, 2

C              ---Test Total Energy

               Call Toterg(Ent(BoxID), Virt(BoxID), BoxID)
               If (Abs(Ent(BoxID)-En(BoxID)).Gt.1.D-6) Then
                  Write (6, *) 
     &                    ' ######### Problems Energy ################ '
               End If
               If (Abs(Virt(BoxID)-Vir(BoxID)).Gt.1.D-6) Then
                  Write (6, *) 
     &                    ' ######### Problems Virial ################ '
               End If
               Write (6, 99002) BoxID, Ent(BoxID), En(BoxID), Ent(BoxID) - En(BoxID), 
     &                          Virt(BoxID), Vir(BoxID), Virt(BoxID) - Vir(BoxID)
            End Do
C           ---Calculation Chemical Potential
            Call Init_Chem(2)
         End If
      End Do
      Call Store(21, Dr, Vmax)

 
99001 Format (' Box : ', I3, /, 
     &        ' Total Energy Initial Configuration : ', F12.5, /, 
     &        ' Total Virial Initial Configuration : ', F12.5)
99002 Format (' Box : ', I3, /, ' Total Energy End Of Simulation   : ', 
     &        F12.5, /, '       Running Energy             : ', F12.5, 
     &        /, '       Difference                 :  ', E12.5, /, 
     &        ' Total Virial End Of Simulation   : ', F12.5, /, 
     &        '       Running Virial             : ', F12.5, /, 
     &        '       Difference                 :  ', E12.5)
99003 Format (' Number Of Att. To Displ. A Part.  : ', I10, /, 
     &        ' Success: ', I10, '(= ', F5.2, '%)')
99004 Format (' Number Of Att. To Chan. Volume    : ', I10, /, 
     &        ' Success: ', I10, '(= ', F5.2, '%)')
99005 Format (' Number Of Att. To Exchange Part.  : ', I10, /, 
     &        ' Success: ', I10, '(= ', F5.2, '%)')
      
      Stop
      End
